include/onnxruntime/core/platform/ort_mutex.h

```diff
@@ -187,3 +188,36 @@ std::cv_status OrtCondVar::wait_for(std::unique_lock<OrtMutex>& cond_mutex,
 }
 };  // namespace onnxruntime
 #endif
+
+namespace onnxruntime {
+
+/*
+OrtSpinLock implemented mutex semantic "lock-freely",
+calling thread will not be put to sleep on blocked,
+which reduces cpu usage on context switching.
+*/
+struct OrtSpinLock {
+  void lock() noexcept {
+    bool occupied = false;
+    while (!occupied_.compare_exchange_weak(
+        occupied,
+        true)) {
+      occupied = false;
+      concurrency::SpinPause();  // pause and retry
+    }
+  }
+  bool try_lock() noexcept {
+    bool occupied = false;
+    return occupied_.compare_exchange_weak(
+        occupied,
+        true);
+  }
+  void unlock() noexcept {
+    occupied_.store(false);
```

Comment:
Using memory_order​::​acquire and memory_order​::​release allows for more efficient implementation on some machine architectures. See 
1. https://www.boost.org/doc/libs/1_82_0/libs/atomic/doc/html/atomic/usage_examples.html#boost_atomic.usage_examples.example_spinlock.implementation
2. http://eel.is/c++draft/atomics.order#6
3. https://rigtorp.se/spinlock/

Reply:
Done.